doc: boards: secure and non-secure domain defs (#9017)
nrfconnect/sdk-nrf: doc/nrf/app_boards.rst

The following tables list all boards and build targets for Nordic Semiconductor's hardware platforms.

> The build target column uses the following terms for multi-core hardware platform entries:

Author: Rephrased into more bullet points. Please recheck after my push.

> * ``cpuapp`` -- When you choose this target, you build the application core firmware that is entirely in the Secure Processing Environment (SPE) and does not use :ref:`Trusted Firmware-M (TF-M) <ug_tfm>`.

Reviewer: Secure Processing Environment as a concept is related to the usage of ArmV8-M Security Extension. I don't think it is completely accurate to call this Secure Processing Environment (SPE)

```suggestion
* ``cpuapp`` -- When you choose this target, you build the application core firmware in a single execution environment, that does not use :ref:`Trusted Firmware-M (TF-M) <ug_tfm>`.
```

Author: Rephrased to "* ``cpuapp`` -- When you choose this target, you build the application core firmware as a single execution environment that does not use Cortex-M Security Extensions (:ref:`Trusted Firmware-M (TF-M) <ug_tfm>`)."

> * ``cpuapp_ns`` -- When you choose this target, you build the application core firmware that is separated between the Secure Processing Environment (SPE) and an additional Non-Secure Processing Environment (NSPE).

Reviewer: I concur. I would rather have this described more accurately describing that the application is in Non-Secure Processing environment (NSPE), and that this relies Secure Processing Environment for security features

> * ``cpuapp_ns`` -- When you choose this target, you build the application core firmware that is separated between the Secure Processing Environment (SPE) and an additional Non-Secure Processing Environment (NSPE).

Author: Rephrased to:
```
* ``cpuapp`` -- When you choose this target, you build the application core firmware that is entirely in the Secure Processing Environment (SPE) and does not use Cortex-M Security Extensions (:ref:`Trusted Firmware-M (TF-M) <ug_tfm>`).
* ``cpuapp_ns`` -- When you choose this target, you build the application with Cortex-M Security Extensions (CMSE).
  The application core firmware is placed in Non-Secure Processing Environment (NSPE) and uses Secure Processing Environment (SPE) for security features.
  The build system automatically includes :ref:`Trusted Firmware-M (TF-M) <ug_tfm>` in SPE and merges it with NSPE.
```

> * ``cpuapp_ns`` -- When you choose this target, you build the application core firmware that is separated between the Secure Processing Environment (SPE) and an additional Non-Secure Processing Environment (NSPE).

Author: This is now rephrased with earlier comments from Håkon and Frank Audun.

  This separation makes this platform more secure than the one without ``_ns``.
* ``cpunet`` -- This indicates the build target for the network core.

> Read more about separation of domains in the :ref:`app_boards_secure_ns_domains` section.

Reviewer 2: Much cleaner solution :)

|                   |            |                                                                 |                                       |
|                   |            |                                                                 | ``thingy53_nrf5340_cpuapp``           |
|                   |            |                                                                 |                                       |

> |                   |            |                                                                 | ``thingy53_nrf5340_cpuapp_ns``*       |

Reviewer 2: The build does not seem to like the asterisk at this spot. Would it work adding a backslash just before the asterisk, like this \*?
Or enclosing it in double back-quotes (inline literals), like this ``*``.

Do we need this though? It feels a bit like a footnote :)

One of the |NCS| applications that lets you add custom boards is :ref:`nrf_desktop`.
See :ref:`nrf_desktop_porting_guide` in the application documentation for details.

.. _app_boards_secure_ns_domains:

Secure and non-secure domains
*****************************

> Secure domain and non-secure domain are two concepts related to the way of organizing multi-core SoC's firmware for the application core, when it is programmed to the ``cpuapp`` or ``cpuapp_ns`` build targets.

Reviewer 2: ```suggestion
Secure domain and non-secure domain are two concepts related to the way of organizing the firmware in the application core of a multi-core SoC, when the firmware is programmed to the ``cpuapp`` or ``cpuapp_ns`` build targets.
```

The firmware for the application core can be assigned to either Secure Processing Environment (SPE) or Non-Secure Processing Environment (NSPE).
Both concepts are related to the security by separation architecture principle, where separating the firmware lets you better protect sensitive assets and code:

* The SPE corresponds to the secure domain and stores all of the security features, including cryptography components and keys.
* The NSPE corresponds to the non-secure domain and stores the main application, among other major components.

In Zephyr and the |NCS|, depending on the build target, you can build the firmware for the application core either for the secure domain alone (build targets without ``_ns``) or for both the secure domain and the non-secure domain at the same time (build targets that include ``_ns``).

> On selecting the build target for the non-secure domain (``_ns``), the application is built as a non-secure image, and an additional firmware component is built as the secure image.

Reviewer 2: Consider adding a blank line before/splitting to new paragraph.

The build system merges both images to form a combined image that will be used for programming or updating the device.
The additional component is :ref:`Trusted Firmware-M (TF-M) <ug_tfm>`.

> (The Secure Partition Manager (SPM) is deprecated.)

Reviewer 2: If we would like to have this information here about the SPM deprecation, then maybe remove the brackets.

One of the |NCS| applications that lets you add custom boards is :ref:`nrf_desktop`.
See :ref:`nrf_desktop_porting_guide` in the application documentation for details.

.. _app_boards_spe_nspe:

Processing environments
***********************

> The build target column in the tables above lists entries ending in ``cpuapp``, ``cpuapp_ns``, and ``_ns`` in the multi-core SoC rows.

Author: Addressed:
"The build target column in the tables above separates entries according to the CPU to target (for multi-core SoCs) and whether Cortex-M Security Extensions (CMSE) are used or not (addition of ``_ns`` if they are used)."

> The ``cpuapp_ns`` and ``_ns`` suffixes in the board target names refer to the adoption of the Security by Separation architecture principle, which splits the firmware to better protect sensitive assets and code.

Author: Indeed, with your previous comment, I see where the confusion was. This is now rephrased to "The ``_ns`` suffix in the board target name refers to the usage of Cortex-M Security Extensions (CMSE)."

> In Security by Separation, the firmware is stored in one of two security domains (partitions), either Secure Processing Environment (SPE) or Non-Secure Processing Environment (NSPE).

Reviewer: Concur: The Secure Processing Environment (and the Arm PSA spec) is already using the term partition when it comes to separation e.g. of ARoT services and PRoT services as well as select services inside the Platform Root of Trust (e.g. PSA crypto service, PSA attestation service and PSA ITS) as being "partitions"

Environment is much better!

> In Security by Separation, the firmware is stored in one of two security domains (partitions), either Secure Processing Environment (SPE) or Non-Secure Processing Environment (NSPE).

Reviewer 2: Oooh, I like the second one here, it makes sense. But don't drop the comma ;)

> In Security by Separation, the firmware is stored in one of two security domains (partitions), either Secure Processing Environment (SPE) or Non-Secure Processing Environment (NSPE).

Author: Rephrased to "two security environments (flash partitions)"

> In Security by Separation, the firmware is stored in one of two security domains (partitions), either Secure Processing Environment (SPE) or Non-Secure Processing Environment (NSPE).

Author: Rephrased to "With CMSE, the firmware is stored in one of two security environments (flash partitions), either Secure Processing Environment (SPE) or Non-Secure Processing Environment (NSPE)."

This isolation of firmware is only possible if the underlying hardware supports `ARM TrustZone`_.

.. figure:: images/spe_nspe.svg
   :alt: Processing environments in the |NCS|

   Processing environments in the |NCS|

In Zephyr and the |NCS|, SPE and NSPE are used exclusively in the context of the application core of a multi-core SoC.
Building follows the Security by Separation principle and depends on the build target.

.. _app_boards_spe_nspe_cpuapp:

> Building for ``cpuapp`` (SPE-only)

Author: This is now rephrased following the suggestion of adding CMSE as the differentiator. I hope this will now read better.

==================================

When you build for ``cpuapp``, you build the firmware for the application core for SPE alone.
There is no separation of firmware and SPE is the only processing environment for the entire application core firmware.
TF-M is not used.

.. _app_boards_spe_nspe_cpuapp_ns:

Building for ``cpuapp_ns`` or ``_ns`` (SPE with NSPE)
=====================================================

When you build for ``cpuapp_ns`` or ``_ns``, you build the firmware for both SPE and NSPE at the same time.

> The firmware is separated between the two environments:

Author: Rephrased to "CMSE is enabled and there is the following separation between the two environments:".

> The firmware is separated between the two environments:

Author: Addressed.

> * SPE implements security-critical functionality and data and isolates them from the application software.

Reviewer: I like the addition. It reminds the user of this detail in clear terms

  SPE contains secure firmware running in the secure state.

> * NSPE implements the main user application and communication firmware, among other major components.

Reviewer: I don't think that we should talk about security-critical code in the NSPE, but the typically is better as it shows that you can do multiple things in that scope.

I propose we make this change

The API also coordinates the communication with components in NSPE.

SPE with ``cpuapp``
-------------------

In the scenario without separation of firmware (``cpuapp``), SPE is the only processing environment for the entire application core firmware.
When you build for such targets, the firmware is placed only in the SPE, without TF-M.

Non-Secure Processing Environment (NSPE)
========================================

NSPE implements the main user application and communication firmware, among other major components.

SPE and NSPE in Zephyr and the |NCS|
====================================

In Zephyr and the |NCS|, SPE and NSPE are used when talking about the application core of a multi-core SoC, when the firmware is programmed to the ``cpuapp`` or ``cpuapp_ns`` build targets.
Depending on the build target, you can build the firmware for the application core either for SPE alone (build targets without ``_ns``) or for both SPE and NSPE at the same time (build targets that include ``_ns``).

> On selecting the build target for NSPE (``_ns``), the application is built as a non-secure image, and an additional firmware component is built as the secure image.

Reviewer 2: "On selecting the build target for NSPE" is contradicting the sentence just before a little bit, in my opinion. I think we might just write "On selecting the ``_ns`` build target" and leave the rest of the sentence as is. Or:
```suggestion
On selecting the ``_ns`` build target, the application is built as a non-secure image to reside in NSPE, and an additional firmware component is built as the secure image to reside in SPE.
```

Author: This section was re-written.

Reviewer 2: Perfect!
